File: DSAlgorithms/Recursion/cpp/Recursion.cpp
#include <iostream>

int numberInLine(int);
int sumOfNumbers(int);
std::string stringReversal(std::string);
bool palindrome(std::string);
int binarySearch(int arr[], int left, int right, int x);
long int fib(long int n);
int main()
{
    std::cout<<numberInLine(5)<<'\n';
    std::cout<<sumOfNumbers(5)<<'\n';
    std::cout<<stringReversal("hello")<<'\n';
    std::cout<<palindrome("sarah")<<'\n';
    std::cout<<palindrome("hannah")<<'\n';
    int sortedNums[] = {1, 0, 3, 4, 6, 7, 8, 9, 11, 16, 17};
    std::cout<<binarySearch(sortedNums, 0, 10, 9)<<'\n';
    std::cout<<fib(5)<<'\n';
    return 0;
}
//simulates being last person waiting in line
int numberInLine(int num)
{
    if(num <= 1)
        return 1;
    return numberInLine(num - 1) + 1;
}
//returns sum of numbers up to n
int sumOfNumbers(int n)
{
    if(n <= 1)
        return 1;
    return sumOfNumbers(n - 1) + n;
}
//reverses a string
std::string stringReversal(std::string input)
{
    if(input.size() == 0)
        return "";
    return stringReversal(input.substr(1, std::string::npos)) + input[0];
}
bool palindrome(std::string input)
{
    if(input.size() == 0 || input.size() == 1)
        return true;
    if(input[0] == input[input.size() - 1])
        return palindrome(input.substr(1, input.size() - 2));
    return false;
}
int binarySearch(int arr[], int left, int right, int x)
{
    if(left > right)
        return -1;
    int mid = (left + right) / 2;
    if(arr[mid] == x)
        return mid;
    if(arr[mid] > x)
        return binarySearch(arr, left, mid - 1, x);
    return binarySearch(arr, mid + 1, right, x);
}
long int fib(long int n)
{
    if(n <= 2)
        return 1;
    return fib(n - 1) + fib(n - 2);
}

File: DSAlgorithms/Recursion/ts/src/Recursion.ts
const numberInLine = (num: number): number => {
    if(num <= 1)
        return 1;
    return numberInLine(num - 1) + 1;
}

const sumOfNumbers = (n: number): number => {
    if(n <= 1)
        return 1;
    return sumOfNumbers(n - 1) + n;
}

const stringReversal = (input: string): string => {
    if(input.length === 0)
        return "";

    return stringReversal(input.substring(1)) + input[0];
}

const isPalindrome = (input: string): boolean => {
    if(input.length === 0 || input.length === 1)
        return true;
    if(input[0] === input[input.length - 1])
        return isPalindrome(input.substring(1, input.length - 1));
    return false;
}
const binarySearch = (Arr: number[], left: number, right: number, x: number): number => {
    if(left > right)
        return -1;
    let mid = Math.floor((left + right) / 2);
    if(x === Arr[mid])
        return mid;
    if(x < Arr[mid])
        return binarySearch(Arr, left, mid - 1, x);
    return binarySearch(Arr, mid + 1, right, x);
}
const sortedArr = [-1, 0, 3, 5, 7, 9, 11, 15];
console.log(numberInLine(5));
console.log(sumOfNumbers(5));
console.log(stringReversal("hello"));
console.log(isPalindrome("sarah"));
console.log(isPalindrome("hannah"));
console.log(binarySearch(sortedArr,0, 7, 11));
